### ISSUE TITLE
Solr pods never become ready when creating a new SolrCloud

### ISSUE THREAD

**Issue by timterle** [bloomberg/solr-operator#46]
**Describe the bug**
When creating a SolrCloud the example-solrcloud-0 pod never becomes ready.

**To Reproduce**
Create a simple SolrCloud example with operator args:
- -zk-operator=true
- -etcd-operator=false

In the logs of the pod example-solrcloud-0 this is found:

```
java.lang.IllegalArgumentException: A HostProvider may not be empty!
	at org.apache.zookeeper.client.StaticHostProvider.<init>(StaticHostProvider.java:88)
	at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:449)
	at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:383)
```

kubectl shows this:
`pod/example-solrcloud-0                             0/1     Running             0          3m`

The solr pod never becomes ready. When checking the pod it says ZK_HOST is "/". When watching the solr statefulset it looks like the pod template has a correct ZK_HOST. But currentRevision and updateRevision for the solr statefulset differ so it's likely that the first revision of the statefulset has an invalid ZK_HOST and the second revision has a correct one. But the pod is stuck with the first revision of the statefulset and can't recover. Kubernetes tries to restart the pod many times, but it's always the same error because ZK_HOST is still "/".

A workaround is to delete the example-solrcloud-0 pod, the installation will then continue and succeed.

**Comment by sepulworld** [comment]
Ah interesting, I think I’ve seen this same race condition as well. Does it happen to you every single time you setup a new SolrCloud? I never dug into why, but your debugging seems to have found the root cause! Nice
    	
    	

    	Get Outlook for iOS
    
  




On Mon, Nov 4, 2019 at 7:55 AM -0800, "timterle" <notifications@github.com> wrote:












Describe the bug

When creating a SolrCloud the example-solrcloud-0 pod never becomes ready.


To Reproduce

Create a simple SolrCloud example with operator args:

-zk-operator=true
-etcd-operator=false



In the logs of the pod example-solrcloud-0 this is found:
java.lang.IllegalArgumentException: A HostProvider may not be empty!
	at org.apache.zookeeper.client.StaticHostProvider.<init>(StaticHostProvider.java:88)
	at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:449)
	at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:383)



kubectl shows this:

pod/example-solrcloud-0                             0/1     Running             0          3m


The solr pod never becomes ready. When checking the pod it says ZK_HOST is "/". When watching the solr statefulset it looks like the pod template has a correct ZK_HOST. But currentRevision and updateRevision for the solr statefulset differ so it's likely that the first revision of the statefulset has an invalid ZK_HOST and the second revision has a correct one. But the pod is stuck with the first revision of the statefulset and can't recover. Kubernetes tries to restart the pod many times, but it's always the same error because ZK_HOST is still "/".


A workaround is to delete the example-solrcloud-0 pod, the installation will then continue and succeed.

**Comment by timterle** [comment]
@sepulworld Yes, actually every time

**Comment by HoustonPutman** [comment]
Fixed in the `v0.2.0` release
